lib/ueberauth/strategy/okta.ex

```diff
@@ -201,7 +201,11 @@ defmodule Ueberauth.Strategy.Okta do
 
   defp add_oauth_options(opts, conn) do
     oauth_opts = Application.get_env(:ueberauth, Ueberauth.Strategy.Okta.OAuth, [])
-    oauth_opts = oauth_opts[strategy_name(conn)] || oauth_opts
-    Keyword.merge(opts, oauth_opts)
+    multitenant_key = conn |> strategy_name() |> String.to_atom()
```

Comment:
My issue is that I don't have an atom corresponding to `strategy_name()`. In fact it's a little weird that a strategy name is required to be passed into `Ueberauth.run_request`, since all relevant options are passed in to the `opts` argument.

Reply:
The documentation actually discourages use of this function for atoms defined outside the current module:

> it is generally recommended to call [String.to_existing_atom/1](https://hexdocs.pm/elixir/String.html#to_existing_atom/1) only to convert atoms defined within the module making the function call to [to_existing_atom/1](https://hexdocs.pm/elixir/String.html#to_existing_atom/1).

https://hexdocs.pm/elixir/String.html#to_existing_atom/1

I can't see an obvious solution here. Maybe it would be better use a map to configure providers

Reply:
Regarding the current strategy for configuration:

```elixir
config :ueberauth, Ueberauth,
  providers: [
    okta: {Ueberauth.Strategy.Okta, []}
  ]

config :ueberauth, Ueberauth.Strategy.Okta.OAuth,
  okta: [
    site: "https://your-doman.okta.com"
    client_id: System.get_env("OKTA_CLIENT_ID"),
    client_secret: System.get_env("OKTA_CLIENT_SECRET")
  ]
```

I wonder why we wouldn't simply define a second Ueberauth provider?

```elixir
config :ueberauth, Ueberauth,
  providers: [
    okta_one:
      {Ueberauth.Strategy.Okta,
       [
         site: "https://your-doman.okta.com",
         client_id: System.get_env("OKTA_CLIENT_ID"),
         client_secret: System.get_env("OKTA_CLIENT_SECRET")
       ]}
    okta_two:
      {Ueberauth.Strategy.Okta,
       [
         site: "https://your-doman.okta.com",
         client_id: System.get_env("OKTA_CLIENT_ID"),
         client_secret: System.get_env("OKTA_CLIENT_SECRET")
       ]}
  ]
```

Reply:
The second option does work _only if_ you have those environment variables at compile time. This is because `config :ueberath, Ueberauth, ...` options are evaluated at compile time in `Plug.init/1`. If you try to set those values at runtime, (i.e. via `runtime.exs` etc), then they are ignored. One main problem here is there are too many ways to attempt to get configuration the tall seem to be competing with each other 🤦🏼 

I'm digging in more and hopefully have a solution soon

Reply:
I tried to capture that here - https://github.com/jjcarstens/ueberauth_okta/blob/main/lib/ueberauth/strategy/okta.ex#L17-L20

If it doesn't make sense, then I'd love recommendations to help improve it

Reply:
Have you encountered a real need to define this configuration in runtime.exs? It seems like this is something that Ueberauth proper doesn't really support, so I wonder if it's necessary for a plugin to implement an independent solution? Maybe it would be better to upstream this feature request?

Reply:
The mutlitenant request came from several sources and the runtime.exs need is very common (see #16) - Most users will not define sensitive environment variables until runtime in production.

I confirmed that `Ueberauth` sets the strategy name to whatever the user sets it to in the config:

```elixir
# config :ueberauth, Ueberauth, providers: [test: []]
iex)> Ueberauth.Strategy.Helpers.strategy_name(conn)
:test

# config :ueberauth, Ueberauth, providers: [%{"test" => []}]
iex)> Ueberauth.Strategy.Helpers.strategy_name(conn)
"test"
```

So there seems to be a mismatch of their documentation between saying to use provider names as atoms and as strings (like [run_request](https://hexdocs.pm/ueberauth/Ueberauth.html#run_request/4) as noted)

I think that since you are using `Ueberauth.run_request/4`, there are quite a few things you can do here to easily fix since you have direct access to the function arguments and don't need to rely on application configuration as much:
* Adjust your `proivder_name` to be an atom before sending to `Ueberauth.run_request/4`
* Adjust your Application config to use a stringified provider name `"okta"`
* Provide the whole `provider_config` details that you need directly to `run_request` and reduce the need of relying on `Ueberauth` or `Ueberauth.Okta` to build it for you

Reply:
This is tricky. Yes, I can work around the issue by converting my strategy name to an atom, but this:
1. Violates the function spec for `strategy_name(conn)`.
2. Feels like a rather brittle workaround, relying on the current internal implementation, rather than a proper solution.

How about simply catching the `ArgumentError` if the atom doesn't already exist?

```elixir
  defp add_oauth_options(opts, conn) do
    oauth_opts = Application.get_env(:ueberauth, Ueberauth.Strategy.Okta.OAuth, [])

    multitenant_key =
      try do
        conn |> strategy_name() |> String.to_existing_atom()
      rescue
        ArgumentError -> nil
      end

    oauth_opts = oauth_opts[multitenant_key] || oauth_opts
    Keyword.merge(opts, oauth_opts)
  end
```

Reply:
Their spec needs to be updated. Ueberauth themselves are describing the the provider name is stored as an atom _and_ also use cases with a stringified name. In fact, all `strategy_name/1` is doing is [pulling a key from the private opts of the config unchanged](https://github.com/ueberauth/ueberauth/blob/master/lib/ueberauth/strategies/helpers.ex#L205-L206). Thats something that should probably be addressed upstream

What we could is just shift this lookup to be indifferent by doing a comparison of both keys as strings

```elixir
defmodule Test do
  def indifferent_get(opts, key) do
    Enum.find_value(opts, opts, fn {k, v} ->
      if to_string(k) == to_string(key), do: v 
    end)
  end
end

iex(1)> opts = [{"howdy", 1}, {:partner, 2}]
[{"howdy", 1}, {:partner, 2}]

iex(3)> Test.indifferent_get opts, :howdy
1
iex(4)> Test.indifferent_get opts, "howdy"
1
iex(5)> Test.indifferent_get opts, "partner"
2
iex(6)> Test.indifferent_get opts, :partner 
2
iex(8)> Test.indifferent_get opts, :no_exist
[{"howdy", 1}, {:partner, 2}]
```

No catching, potential atom leaks, or caring about the internal helpers of ueberauth. What do you think?

Reply:
I do highly suggest to be in charge of your own options as well. `run_request` seems to be for that exact purpose of having more explicit control and _not_ relying on internals or some lib to manage your options. Though the above is also and acceptable convenience

Reply:
> Their spec needs to be updated.

Agreed :) But since the spec is unclear, I think it's better not to rely too much on the current implementation: they may well solve it by always converting `provider_name` to a string, rather than the other way.

> What we could is just shift this lookup to be indifferent by doing a comparison of both keys as strings
> No catching, potential atom leaks, or caring about the internal helpers of ueberauth. What do you think?

Yeah, if you like that better, that works just fine for me :+1: 

> I do highly suggest to be in charge of your own options as well. `run_request` seems to be for that exact purpose of having more explicit control and _not_ relying on internals or some lib to manage your options. Though the above is also and acceptable convenience

Yes, absolutely - the reason I'm using `run_request` is precisely because I store the options elsewhere in config, and have some additional logic that I'm applying to them. I am not using the build-in config mechanisms at all, which I think is what is tripping up this code. That's why there is no atom corresponding to the `provider_name` string, for instance.

Reply:
@jjcarstens Any thoughts on whether you'd like to go with the `ArgumentError rescue` approach or implement your `indifferent_get` approach? The first one is what is ready currently in this PR.

Reply:
@giddie So sorry! I got really behind on this.

I think I'd rather do the indifferent get based on strings rather than do any atom conversion attempts. That should cover both atom or string key cases